### ISSUE TITLE
State Change in onSubmit Breaks Validation

### ISSUE THREAD

**Issue by devinsm** [rjsf-team/react-jsonschema-form#2408]
### Prerequisites

- [X] I have read the [documentation](https://react-jsonschema-form.readthedocs.io/);
- [X] In the case of a bug report, I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [X] Ideally, I'm providing a [sample JSFiddle](https://jsfiddle.net/n1k0/f2y3fq7L/6/) or a [shared playground link](https://rjsf-team.github.io/react-jsonschema-form/) demonstrating the issue.

### Description

I have created a [minimal reproduction](https://codesandbox.io/s/github/devinsm/rjsf-validation-bug) (the reproduction is also [available on GitHub](https://github.com/devinsm/rjsf-validation-bug)).

The bug: if you use the `Form` component as a controlled component, and state changes in the `onSubmit` handler, form validation doesn't work right. When validation fails, it will continue to fail even if the problems have been corrected.

### Steps to Reproduce

Run the reproduction and:
1. Enter 0 in "Magic Number" input (the form is now invalid)
1. Click "Submit" (it will fail since the form is invalid)
1. Change "Magic Number" back to 2 (the form is now valid)
1. Hit "Submit" again... no matter how many times you click it the UI still
   shows an error

#### Expected behavior

Once you fix validation errors you should be able to submit the form and the UI should no longer display the errors.

#### Actual behavior

If the form fails to validate when the user first submits the form, then after correcting the errors form validation will continue to fail.

### Version

`2.5.1`

**Comment by AlexMachin1997** [comment]
I think I'm having a similar issue on my work project.

It's really anoying, it's probably the fact that the Form component uses an unsafe lifecylce method e.g. componentWillRecieveProps. 

**What happens in my project:**

Submit an invalid form submission

Invalid error messages are present

Set form data, it's now valid.

In the onSubmit I set the formData via the useState hook, it still shows the error messages even though it's a valid payload.

**Note's:**

Honeslty I'm not impressed with how this library work's, it's awful. It's just prop drilling (I've overriden alot of the default functionality of the form so I know how bad it is). It need's to be better, in this particular case it should read the data in once and then don't attempt to re-read it. I'm only really setting the form data in the onSubmit because when I put the form into submitting mode via isSubmitting state in my app the formData reset as that unsafe lifecylce is being run.

**Comment by epicfaace** [comment]
This is a related PR which _might_ fix this: https://github.com/rjsf-team/react-jsonschema-form/pull/2010

**Comment by devinsm** [comment]
@epicfaace I could add a test to the test suite if you'd like.

**Comment by AlexMachin1997** [comment]
@epicfaace Look's like it's been open since 2020, I won't hold my breath.

I can see it uses the componentDidMount to set the inital form data for situations such as editing, but it still has the componentDidUpdate which still triggers the formData update from what I gathered. You shouldn't need that or the deepEquals stuff.

**Comment by devinsm** [comment]
@AlexMachin1997 I like how invested you are in this issue, and appreciate that you've dug into the code to try to pinpoint the source of the bug. However I think it's important to give @epicfaace and team credit for all the hard work they've put into this package.

It may not be perfect, and it does have bugs, but then again that's true of most software. Looking through the issues and commits, I see a team of open source developers that really does care and really is trying to make things better. They clearly invested hundreds if not thousands of hours with no expectation of financial gain.

When you say things like "look's like it's been open since 2020, I won't hold my breath" or "honestly I'm not impressed with how this library work's, it's awful", I can only imagine how it makes the maintainers feel. Remember that for them this is a work of love. Would you say those things to an artist at a local gallery?

Again, I appreciate your help with this issue. I hope we can work together to continue to make GitHub a welcoming and supportive community.

**Comment by AlexMachin1997** [comment]
@devinsm I'm not after any credit hahah. 

I'm just happy to see other people are having this issue and it's not just me 😄 

Sorry for my comment, I don't usually post comment's like that I'm usually a postive person and offer help where needed.

This library is so powerful and can do so much 💯 It's amazing what the form can be done out of the box, and when you do need something custom you can custom it.

**Comment by epicfaace** [comment]
Thanks for the note @devinsm , appreciate it. And no worries, @AlexMachin1997.

For additional context, I began maintaining this library in 2019. Before that, it was created and owned by Mozilla, but it had really fallen into disuse -- no one was really merging PRs and releases were rarely happening. I just contacted the maintainers and offered to help out. Eventually, we moved it over to the rjsf-team organization, as the rjsf maintainers at Mozilla were just happy to have someone else (me) maintain the project (see https://github.com/rjsf-team/react-jsonschema-form/issues/1440).

Over time, rjsf has just become one of a dozen other projects I'm working on, and it's unpaid, so I really only have limited bandwidth to work on this project (maybe merging a PR or two a week, keeping releases going, etc.). There are three solutions to this:
- Get more maintainers, i.e., people who can review PRs and provide good feedback to ensure the code base has good code. See https://github.com/rjsf-team/react-jsonschema-form/issues/1757. A couple people have signed up to help (and have been great and super helpful!) but no one other than me (perhaps understandably) has yet had the time to consistently review the PRs that keep coming in to the project. If you're interested in helping, let me know!
- Make it easier for me to work on your issue by sponsoring me! Obviously, I'm still maintaining this library for free and will likely get to major bug fixes / issues eventually, but if you / your company would like me to put in more effort than just what I can manage in my free time, the most sustainable way would be to sponsor me: https://github.com/sponsors/epicfaace/
